### PR TITLE
swarm/storage: Write every batch on iteration

### PR DESCRIPTION
**Pull request by nolash**
Currently `CleanGCIndex` stores all batches in memory. That's bad news for any large database.

### CODE REVIEW

File: swarm/storage/ldbstore.go [hunk 1]
```diff
@@ -633,6 +633,9 @@ func (s *LDBStore) CleanGCIndex() error {
 			}
 		}
 		totalEntryCount++
+		if err := s.db.Write(&batch); err != nil {
```
Comment (nonsense) [inline swarm/storage/ldbstore.go]: If you add `batch := leveldb.Batch{}` on line 608, that might work.

File: swarm/storage/ldbstore.go [hunk 2]
```diff
@@ -602,41 +602,76 @@ func (s *LDBStore) CleanGCIndex() error {
...
+	it.Release()
+
+	//var idx dpaDBIndex
 	var poPtrs [256]uint64
-	for it.Valid() {
-		rowType, chunkHash := parseGCIdxKey(it.Key())
-		if rowType != keyIndex {
-			break
+	var doneIterating bool
+	for !doneIterating {
```
Comment (nonsense) [inline swarm/storage/ldbstore.go:621]: We should add a `log.Debug` line here to use as a progress indicator when we run this.

**Review by frncmx — COMMENTED** [review]
- Is there a specific reason you are using the `err := callFunc(); if err != nil ...` format instead of `if err := callFunc(); err != nil...`?
- I find `cleanGCIndex` too long; it does not fit into my head. It's more than a 100 lines with quite a code complexity. Could we split it up into higher level functions?

**Comment by nolash** [comment]
> I find cleanGCIndex too long; it does not fit into my head. It's more than a 100 lines with quite a code complexity. Could we split it up into higher level functions?

Sure if you think that's better. Any suggestions on how and what?

File: swarm/storage/ldbstore.go [hunk 3]
```diff
@@ -602,43 +602,87 @@ func (s *LDBStore) CleanGCIndex() error {
...
+		batch.Reset()
+		it.Seek(lastIdxKey)
+
+		var i int
+		for i = 0; i < 4096; i++ {
+			if !it.Valid() {
+				doneIterating = true
+				break
+			}
+			rowType, chunkHash := parseGCIdxKey(it.Key())
```
[the remark this answers is not in the document]
Comment (nolash) [reply swarm/storage/ldbstore.go]: Thanks. I didn't notice the name discrepancy here, and I don't really recognize this `parseGCIdxKey` function. Weird.

Looking at what it does, though, it's generic across all Indices. We could either just call it `parseIdxKey` or we can add `parseKeyIdxKey` and replace this with that.

[the remark this answers is not in the document]
Comment (nolash) [reply swarm/storage/ldbstore.go]:
```
c41e1bd1eb (Ferenc Szabo      2018-11-13 15:22:53 +0100  287) func parseGCIdxKey(key []byte) (byte, []byte) {
c41e1bd1eb (Ferenc Szabo      2018-11-13 15:22:53 +0100  288)   return key[0], key[1:]
c41e1bd1eb (Ferenc Szabo      2018-11-13 15:22:53 +0100  289) }
c41e1bd1eb (Ferenc Szabo      2018-11-13 15:22:53 +0100  290) 
```

Aha, so it's new. That's why. Good, let's just rename then maybe?

**Comment by frncmx** [comment]
> > I find cleanGCIndex too long; it does not fit into my head. It's more than a 100 lines with quite a code complexity. Could we split it up into higher level functions?
> 
> Sure if you think that's better. Any suggestions on how and what?

I would try to capture the main things we are doing here.
- delete all GCIndex entries
- create two piles => bad and good
- delete bad
- write good indexes
- I don't really understand the last thing (`keyDistanceCnt`)

To achieve these separate functions I think it's okay to create new iterators and don't pass (reset) around iterators. (I'm not on expert on LevelDB in this regards... but we have to seek anyway.)

**Comment by nolash** [comment]
> create two piles => bad and good

Isn't this unnecessary? Each has one operation.

**Comment by nolash** [comment]
But yes, then maybe:

purgeGCIndex() {
    remove all gc indices
}

cleanDbIndex() {
    remove orphaned indices
    add gc indices for valid data
}

cleanDbSummary() {
    update new correct indexcount
    update po index ptrs
}

or something?

**Comment by nolash** [comment]
`keyDistanceCnt` explained:

A store is specific to a specific bzzaccount.

All chunks are stored with a `po` relative to that account

The index of the _last_ chunk stored with a specific `po` is recorded in the index with key `keyDistanceCnt`

If no chunks have been added with a particular `po`, there will be no entry for key `keyDistanceCnt` for that `po`.

The name is really misleading. It's not a `C`ou`nt`.
Ok?

File: swarm/storage/ldbstore.go [hunk 4]
```diff
@@ -601,47 +601,113 @@ func (s *LDBStore) CleanGCIndex() error {
...
+				return fmt.Errorf("corrupt index: %v", err)
+			}
+			po := s.po(chunkHash)
+			lastIdxKey = it.Key()
+
+			// if we don't find the data key, remove the entry
+			// if we find it, add to the array of new gc indices to create
+			dataKey := getDataKey(idx.Idx, po)
+			_, err = s.db.Get(dataKey)
+			if err != nil {
```
Comment (frncmx) [inline swarm/storage/ldbstore.go:661]: > > create two piles => bad and good
> 
> Isn't this unnecessary? Each has one operation.

I thought we are kinda doing that already. We collect then write the good indexes only later.
The difference is that we immediately delete bad ones. That breaks symmetry for me.

Anyhow. I did not do a change request on purpose; just a comments. So it's not necessary.

**Comment by frncmx** [comment]
`cleanDbSummary()` is the `keyDistanceCnt` you call summary?
I guess we also have `rebuildGCIndex()`.

**Comment by zelig** [comment]
lgtm to submit upstream

**Review by acud — APPROVED** [review]
lgtm

**Pull request closed without merging**